src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs

```diff
@@ -50,6 +51,67 @@ void AppendLines (string prefix, List<string> lines, StringBuilder sb)
 			}
 		}
 
+		public static int RunProcess (string command, string arguments, TaskLoggingHelper log, DataReceivedEventHandler? onOutput = null, DataReceivedEventHandler? onError = null)
+		{
+			var stdout_completed = new ManualResetEvent (false);
+			var stderr_completed = new ManualResetEvent (false);
+			var psi = new ProcessStartInfo () {
+				FileName = command,
+				Arguments = arguments,
+				UseShellExecute = false,
+				RedirectStandardOutput = true,
+				RedirectStandardError = true,
+				CreateNoWindow = true,
+				WindowStyle = ProcessWindowStyle.Hidden,
+			};
+
+			var stdoutLines = new List<string> ();
+			var stderrLines = new List<string> ();
+
+			log.LogDebugMessage ($"Running process: {psi.FileName} {psi.Arguments}");
+			using var proc = new Process ();
+			proc.OutputDataReceived += (s, e) => {
+				if (e.Data != null) {
+					onOutput?.Invoke (s, e);
+					stdoutLines.Add (e.Data);
+				} else
+					stdout_completed.Set ();
```

Comment:
`e.Data` is null when the process exits?!  I was not aware of that, nor do I see that documented at [`Process.OutputDataReceived`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.outputdatareceived?view=net-8.0).

The following code also feels odd, as it suggests that events may still be raised *after* `proc.WaitForExit()`, hence the need for e.g. `stderr_completed.WaitOne()`, which just feels odd.

Reply:
This pattern was established some 10-11 years ago, we've seen all these conditions happen along the years (especially when I worked with the Xamarin Installer, but also in Mono Develop/Xamarin Studio).  `e.Data == null` is [documented here](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.datareceivedeventargs.data?view=net-8.0#remarks):

```
When the redirected stream is closed, a null line is sent to the event handler. Ensure your event handler checks the [Data](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.datareceivedeventargs.data?view=net-8.0) 
property appropriately before accessing it. For example, you can use the static method [String.IsNullOrEmpty](https://learn.microsoft.com/en-us/dotnet/api/system.string.isnullorempty?view=net-8.0) to validate 
the [Data](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.datareceivedeventargs.data?view=net-8.0) property in your event handler.
```

Process exiting means only that the process itself is gone, its output streams may still have data that needs to be piped to us, hence the need to wait.